go/vt/sqlparser/ast_format.go

```diff
@@ -1361,6 +1361,53 @@ func (node *Literal) Format(buf *TrackedBuffer) {
 
 // Format formats the node.
 func (node *Argument) Format(buf *TrackedBuffer) {
+	// We need to make sure that any value used still returns
+	// the right type when interpolated. For example, if we have a
+	// decimal type with 0 scale, we don't want it to be interpreted
+	// as an integer after interpolation as that would the default
+	// literal interpretation in MySQL.
+	switch {
+	case node.Type == sqltypes.Unknown:
+		// Ensure we handle unknown first as we don't want to treat
+		// the type as a bitmask for the further tests.
+		// do nothing, the default literal will be correct.
+	case sqltypes.IsDecimal(node.Type):
+		buf.astPrintf(node, "CAST(:%#s AS DECIMAL(%d, %d))", node.Name, node.Size, node.Scale)
+		return
+	case sqltypes.IsUnsigned(node.Type):
+		buf.astPrintf(node, "CAST(:%#s AS UNSIGNED)", node.Name)
+		return
+	case node.Type == sqltypes.Float64:
+		buf.astPrintf(node, "CAST(:%#s AS DOUBLE)", node.Name)
+		return
+	case node.Type == sqltypes.Float32:
+		buf.astPrintf(node, "CAST(:%#s AS FLOAT)", node.Name)
+		return
+	case sqltypes.IsDate(node.Type):
+		if node.Size == 0 {
+			buf.astPrintf(node, "CAST(:%#s AS DATE)", node.Name)
+			return
+		}
+		buf.astPrintf(node, "CAST(:%#s AS DATE(%d))", node.Name, node.Size)
```

Comment:
`DATE` can't have sub second precision, in fact MySQL won't parse that.

Reply:
Fixed via [35bdac5](https://github.com/vitessio/vitess/pull/16232/commits/35bdac5544ec5c1034ca45ac93d1de88436b6d30)